### ISSUE TITLE
bsc does not update generated verilog/sim when BSV preprocessor macros changed

### ISSUE THREAD

**Issue by kammoh** [B-Lang-org/bsc#290]
```bsv
package A;

module mkA();
`ifdef MY_MACRO
// behavior 1
rule behavior_1;
$display("behavior 1");
endrule
`else
// behavior 2
rule behavior_2;
$display("behavior 2");
endrule
`endif
endmodule
endpackage
```

If I first run
```
bsc -verilog -g mkA -D MY_MACRO  A.bsv
```
the generated verilog is based on the code in "behavior 1" block.
But then, if I subsequently run:
```
bsc -u -verilog -g mkA A.bsv
```
The module is not recompiled and the Verilog is still based on "behavior 1" block.  I expect bsc to analyze the module again and re-generate the verilog based on "behavior 2".

I need to manually remove the `.bo` files and rerun bsc to get the expected behavior.

**Comment by quark17** [comment]
Yes, this is something that should be fixed.  This is recorded as Bug 1557 in Bluespec's private bug database (that I am working on moving over to GitHub) and also Bug 97 for the more general case of other flag changes.

The command-line flags would need to be recorded in the `.bo` file (and perhaps also in the `.ba` files), which I don't believe is currently done.  And then the recompile check would need to compare the current and recorded flags and consider any differences.  Any differences in the preprocessor environment should force a recompile; and differences in flags that control synthesis behavior (such as `--agressive-conditions`) should probably also force recompile, when synthesis is specified; but other flag differences probably shouldn't (such as flags to dump info).  Some flags are less clear (such as include paths).  It's possible that we might want to add new flags, to control whether these gray areas should force recompile or not.
